c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,338 @@ WebSocketPipe newWebSocketPipe() {
+  // Removes any leading spaces from `str`, modifying it in-place.
+  while (str.size() > 0 && (str[0] == ' ' || str[0] == '\t')) {
+    str = str.slice(1, str.size());
+  }
+}
+
+kj::Vector<kj::String> splitPartsWithCondition(StringPtr cursor, char delim, StringPtr condition) {
+  // Given a string `input` and a delimiter `delim`, split the string into a vector of substrings,
+  // separated by the delimiter. Note that leading whitespace is stripped from each element.
+  //
+  // Optionally, the caller may pass a `condition` string, requesting that each substring starts
+  // with the value inside `condition`. This allows for filtering in 1 pass, rather than 2.
+  kj::Vector<kj::String> parts;
+
+  while (cursor.size() != 0) {
+    auto part = splitNext(cursor, delim);
+    stripLeadingSpace(part);
+
+    if (part.startsWith(condition.asArray())) {
+      parts.add(kj::str(part));
+    }
+  }
+
+  return parts;
+}
+
+kj::Array<keyMaybeVal> toKeysAndVals(const kj::ArrayPtr<kj::String>& params) {
+  // Given a collection of parameters (a single offer), parse the parameters into <key, MaybeValue>
+  // pairs. If the parameter contains an `=`, we set the `key` to everything before, and the `value`
+  // to everything after. Otherwise, we set the `key` to be the entire parameter.
+  auto result = kj::heapArray<keyMaybeVal>(params.size());
+  size_t count = 0;
+  for (const auto& param : params) {
+    kj::ArrayPtr<const char> key;
+    kj::Maybe<kj::ArrayPtr<const char>> value;
+
+    KJ_IF_MAYBE(index, param.findFirst('=')) {
+      // Found '=' so we have a value.
+      key = param.slice(0, *index);
+      value = param.slice(*index + 1, param.size());
+    } else {
+      key = param.asArray();
+    }
+
+    result[count] = keyMaybeVal{.key = kj::mv(key), .val = kj::mv(value)};
+    ++count;
+  }
+  return kj::mv(result);
+}
+
+struct ParamType {
+  bool isClientParam = false;
+  // `isClientParam` is true if `key` starts with `client`, false otherwise (implies server).
+  bool isCtxTakeover = false;
+  // `isCtxTakeover` is true if `key` is a `no_context_takeover` parameter, false otherwise
+  // (implies max_window_bits).
+};
+
+inline kj::Maybe<ParamType> isValidKeyName(kj::ArrayPtr<const char>& key) {
+  // Returns a `ParamType` struct if the `key` is valid and nullptr if invalid.
+  //
+  // The following `key`s are valid, assume x = "client" or "server".
+  //  - x_no_context_takeover
+  //  - x_max_window_bits
+  // Trailing whitespace is allowed -- any other character invalidates the `key`.
+  const auto client = "client"_kj;
+  const auto server = "server"_kj;
+  const auto ctxTakeover = "_no_context_takeover"_kj;
+  const auto maxBits = "_max_window_bits"_kj;
+
+  ParamType paramType;
+
+  if (key.startsWith(client)) {
+    paramType.isClientParam = true;
+  } else if (!key.startsWith(server)) {
+    // `Key` must start with either "client" or "server" -- invalid.
+    return nullptr;
+  }
+  key = key.slice(client.size(), key.size());
```

Comment:
https://github.com/capnproto/capnproto/pull/1480/commits/a451522fca9207d6bcef083beb7360b956373b39

Had to keep it a bit longer to check trailing characters (since trailing whitespace is allowed).